### PR TITLE
set maximum number of connections to 60000

### PR DESCRIPTION
**Pull request by breuerfelix**
Signed-off-by: Felix Breuer <fbreuer@pm.me>

### CODE REVIEW

File: internal/helper/yawollet.go [hunk 1]
```diff
@@ -344,17 +356,25 @@ func createEnvoyTCPListener(
...
 		FilterChains: []*envoylistener.FilterChain{{
-			// proxy filter has to be the last in the chain
-			Filters: append(filters, &envoylistener.Filter{
-				Name: envoywellknown.TCPProxy,
-				ConfigType: &envoylistener.Filter_TypedConfig{
-					TypedConfig: listenPort,
+			Filters: append(filters,
+				&envoylistener.Filter{
+					// NOTE: this is not present in envoywellknown
+					Name: "extensions.filters.network.connection_limit",
```
Comment (nightlyone) [inline internal/helper/yawollet.go:362]: There is a related constant not mentioned in `envoywellknown`. Maybe you want to put them closed together so you can easily maintain those constants when `envoywellknown` starts adding them.

File: internal/helper/yawollet.go [hunk 2]
```diff
@@ -344,17 +356,25 @@ func createEnvoyTCPListener(
 			},
 		},
 		FilterChains: []*envoylistener.FilterChain{{
-			// proxy filter has to be the last in the chain
-			Filters: append(filters, &envoylistener.Filter{
-				Name: envoywellknown.TCPProxy,
-				ConfigType: &envoylistener.Filter_TypedConfig{
-					TypedConfig: listenPort,
+			Filters: append(filters,
+				&envoylistener.Filter{
```
Comment (nightlyone) [inline internal/helper/yawollet.go:360]: Is there an ordering constraint like "must/should be first in chain?" for this filter? Looks like you want to drop those connections fast before filtering further.

**Review by nightlyone — COMMENTED** [review]
Looks like a very good idea to manage it more explicitly.

File: internal/helper/yawollet.go (hunk 2 above)
Comment (dergeberl) [reply internal/helper/yawollet.go:360]:
```

**Comment by dergeberl** [comment]
We do also limits in the `createEnvoyCluster()` function in `internal/helper/yawollet.go`:
```go
	Thresholds: []*envoycluster.CircuitBreakers_Thresholds{
		{
			Priority:           envoycore.RoutingPriority_DEFAULT,
			MaxConnections:     &wrappers.UInt32Value{Value: 10000 * uint32(hostmetrics.GetCPUNum())},
			MaxRequests:        &wrappers.UInt32Value{Value: 8000 * uint32(hostmetrics.GetCPUNum())},
			MaxPendingRequests: &wrappers.UInt32Value{Value: 2000 * uint32(hostmetrics.GetCPUNum())},
		},
	},
```

This is set per upstream Port which even do not make sense. If I have 10 Ports the sum of all MaxConnections are 100.000 (with 1 CPU). 
Should we set`MaxConnections` here always to a high value (default is 1024) and set it by CPUCount in the new Filter?

But the new setting dedicated for each listener port or is it global. Because it do also not make sense to limit each listener port to 60.000 connections. 
If you run a LoadBalancer with 2 ports you still can get problems with no ports left on the LoadBalancer host.

File: internal/helper/yawollet.go [hunk 3]
```diff
@@ -305,6 +307,16 @@ func createEnvoyTCPListener(
...
 	}
 
+	connectionLimit, err := anypb.New(&envoyconnection.ConnectionLimit{
+		StatPrefix: "envoyconnectionlimit",
+		// this is limited by the number of available ports on the machine
+		MaxConnections: wrapperspb.UInt64(60000),
+	})
+
+	if err != nil {
+		panic(err)
```
Comment (dergeberl) [inline internal/helper/yawollet.go:317]: Maybe just return with error and post a event to the lbm/lb object. A panic inside the yawollet is hard to find and see.

File: internal/helper/yawollet.go (hunk 2 above)
Comment (breuerfelix) [reply internal/helper/yawollet.go:360]: I also thought about that but the only filter before is the rbac filter which is kind of a "security" filter. To me it seems like this one should always be the first in the chain.

What do you think about that? I can still put the filter into its own statement after the if statement, so it is easier to decide for new filters at which position they should go later on.

**Comment by breuerfelix** [comment]
> We do also limits in the `createEnvoyCluster()` function in `internal/helper/yawollet.go`:
> 
> ```go
> 	Thresholds: []*envoycluster.CircuitBreakers_Thresholds{
> 		{
> 			Priority:           envoycore.RoutingPriority_DEFAULT,
> 			MaxConnections:     &wrappers.UInt32Value{Value: 10000 * uint32(hostmetrics.GetCPUNum())},
> 			MaxRequests:        &wrappers.UInt32Value{Value: 8000 * uint32(hostmetrics.GetCPUNum())},
> 			MaxPendingRequests: &wrappers.UInt32Value{Value: 2000 * uint32(hostmetrics.GetCPUNum())},
> 		},
> 	},
> ```
> 
> This is set per upstream Port which even do not make sense. If I have 10 Ports the sum of all MaxConnections are 100.000 (with 1 CPU). Should we set`MaxConnections` here always to a high value (default is 1024) and set it by CPUCount in the new Filter?
> 
> But the new setting dedicated for each listener port or is it global. Because it do also not make sense to limit each listener port to 60.000 connections. If you run a LoadBalancer with 2 ports you still can get problems with no ports left on the LoadBalancer host.

Good point, i haven't found any option which limits the overall connections for envoy across all listeners / clusters yet.  
I will put this PR into draft mode since i requires another architecture.

**Comment by breuerfelix** [comment]
I thought about this and i guess a solid way is to handle this on a system level. We can set the maximum number of open files for the `yawol` user to 60k and create another user which runs yawollet.  
That way the yawollet is still able to open connections to the k8s api, even though we are close to our limits since envoy is not able to open more files.

File: internal/helper/yawollet.go [hunk 4]
```diff
@@ -344,17 +369,17 @@ func createEnvoyTCPListener(
...
 					TypedConfig: listenPort,
 				},
 			}),
 		}},
-		Freebind:                      &wrappers.BoolValue{Value: true},
-		EnableReusePort:               &wrappers.BoolValue{Value: true},
-		PerConnectionBufferLimitBytes: &wrappers.UInt32Value{Value: 32768}, // 32 Kib
+		Freebind:                      wrapperspb.Bool(true),
+		EnableReusePort:               wrapperspb.Bool(true),
+		PerConnectionBufferLimitBytes: wrapperspb.UInt32(32768), // 32 Kib
```
Comment (nightlyone) [inline internal/helper/yawollet.go:382]: You can drop the comment, if you use `32 << 10` or `32 * 1024` here. Then it is clear enough without the comment and doesn't need extra maintenance once you want to change the value.

**Review by dergeberl — APPROVED** [review]
LGTM

**Pull request merged**